modules/howtos/examples/Import.java

```diff
@@ -0,0 +1,411 @@
+        new CSVReaderHeaderAwareBuilder(
+          new FileReader("howtos/examples/import.tsv"))
+        .withCSVParser(parser)
+        .build();
+
+      Map<String, String> row;
+      while ((row = tsv.readMap()) != null) {
+        upsertRow(row);
+      }
+    }
+    // ...
+    // tag::omit[]
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+    catch (com.opencsv.exceptions.CsvValidationException e) {
+      System.out.println("handle CsvValidationException...");
+    }
+    // end::omit[]
+  }
+  // end::importTSV[]
+  
+  // tag::importTSV_batch[]
+  public void importTSV_batch() {
+    // tag::omit[]
+    System.out.println("importTSV_batch");
+    // end::omit[]
+
+    Flux<Map<String,String>> rows = Flux.generate(
+      
+      () -> {
+        CSVParser parser =
+          new CSVParserBuilder()
+          .withSeparator('\t')
+          .withIgnoreQuotations(true)
+          .build();
+        return
+          new CSVReaderHeaderAwareBuilder(
+            new FileReader("howtos/examples/import.tsv"))
+          .withCSVParser(parser)
+          .build();
+      },
+      
+      // ...
+      // tag::omit[]
+      (state, sink) -> {
+        try {
+          Map<String,String> row = state.readMap();
+          if (row == null) {
+            sink.complete();
+          }
+          else {
+            sink.next(row);
+          }
+          return state;
+        }
+        catch (IOException e) { throw new Error("IOException"); }
+        catch (com.opencsv.exceptions.CsvValidationException e) { throw new Error("IOException"); }
+      });
+
+    Flux<MutationResult> results = 
+      rows
+      .map(row -> preprocess(row))
+      .flatMap(doc -> reactiveCollection.upsert(doc.getId(), doc.getContent()));
+      
+    results.subscribe(System.out::println);
+    results.blockLast(Duration.of(60, ChronoUnit.SECONDS));
+
+    System.out.println("DONE");
+    // end::omit[]
+  }
+// end::importTSV_batch[]
+  
+  // tag::importJSON[]
+  public void importJSON() {
+    try {
+      String content  = new String(
+        Files.readAllBytes(
+          Paths.get("howtos/examples/import.json")),
+        StandardCharsets.UTF_8);
+      
+      for (Object row: JsonArray.fromJson(content)) {
+        JsonObject json = ((JsonObject) row);
+        upsertRow(json.toMap());
+      }
+    }
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+  }
+  // end::importJSON[]
+  
+  // tag::importJSON_batch[]
+  public void importJSON_batch() {
+    // tag::omit[]
+    System.out.println("importJSON_batch");
+    // end::omit[]
+
+    try {
+      String content  = new String(
+        Files.readAllBytes(
```

Comment:
NIT: Perhaps a note that this will fail if the whole file doesn't fit in memory? Ideally we'd want to use a JSON stream parser to read each object one by one, but maybe that's out of scope.

Reply:
Good catch. Yes, ideally would be streaming, using the Jackson Streaming API... 

I think this might be for a future version though (will check priority with @simon-dew) unless you happen to have time to contribute the code.